pkg/accounting/accounting.go

```diff
@@ -135,22 +141,70 @@ func (a *Accounting) Credit(peer swarm.Address, price uint64) error {
 
 	a.logger.Tracef("crediting peer %v with price %d, new balance is %d", peer, price, nextBalance)
 
-	err = a.store.Put(peerBalanceKey(peer), nextBalance)
+	// compute expected debt before update because reserve still includes the amount that is deducted from the balance
+	expectedDebt := balance.expectedDebt()
+
+	err = a.saveBalance(peer, nextBalance)
 	if err != nil {
 		return err
 	}
-
 	balance.balance = nextBalance
 
 	a.metrics.TotalCreditedAmount.Add(float64(price))
 	a.metrics.CreditEventsCount.Inc()
 
-	// TODO: try to initiate payment if payment threshold is reached
-	// if balance.balance < -int64(a.paymentThreshold) { }
+	// if our expected debt exceeds our payment threshold (which we assume is also the peers payment threshold), trigger settlement
+	if expectedDebt >= a.paymentThreshold {
+		err = a.settle(peer, balance)
+		if err != nil {
+			a.logger.Errorf("failed to settle with peer %v: %v", peer, err)
+		}
+	}
 
 	return nil
 }
 
+func (a *Accounting) saveBalance(peer swarm.Address, nextBalance int64) error {
+	err := a.store.Put(peerBalanceKey(peer), nextBalance)
+	if err != nil {
+		return fmt.Errorf("failed to persist balance: %w", err)
+	}
+	return nil
```

Comment:
I do not see the need to annotate this error. It would be just fine to return it without the need for the if statement. I would explicitly annotate errors in places where the function is called instead implicitly here.

Reply:
I dropped this function entirely and moved the errorf to where put is called in the individual functions.